scripts/benchmarks/README.md

```diff
@@ -92,3 +92,102 @@ python benchmark.py --help
+
+The HFTrainer API is more granular than `nvidia-smi` as 
+  - It reports the allocated memory by calling `torch.cuda.memory_allocated()` and `torch.cuda.max_memory_allocated()` inside its probes
+  - It has memory logging probes at different stages of the Trainer - `init`, `train`, `evaluate`, `predict` 
+
+##### NOTE:
+When in distributed mode, the Trainer will only log the rank 0 memory.
+
+#### Deciphering the Memory Metrics
+When the memory metrics are logged, additional metrics are included in the output of `Trainer.train / Trainer.evaluate / Trainer.predict`. 
+
+The additional metrics consist of:
+A. absolute memory before trainer initialization
+B. delta of allocated memory of a stage
+C. peaked_delta of a stage (extra mem consumed and freed in between) 
+
+B & C are taken before/after each of the following stages:
+- <init>: `HFTrainer.__init__`
+- <train>: `HFTrainer.train` (only if called)
+- <evaluate>: `HFTrainer.evaluate` (only if called)
+- <predict>: `HFTrainer.predict` (only if called)
+
+Example Output:
+
+```
+output_metrics = {
+    'train_runtime': 143.6704, 
+    'train_samples_per_second': 0.835, 
+    'train_steps_per_second': 0.209, 
+    'train_tokens_per_second': 3421.164, 
+    'train_loss': 1.292790667215983, 
+    'init_mem_cpu_alloc_delta': 8192, 
+    'init_mem_gpu_alloc_delta': 0, 
+    'init_mem_cpu_peaked_delta': 0, 
+    'init_mem_gpu_peaked_delta': 0, 
+    'train_mem_cpu_alloc_delta': 531410944, 
+    'train_mem_gpu_alloc_delta': 126487040, 
+    'train_mem_cpu_peaked_delta': 0, 
+    'train_mem_gpu_peaked_delta': 11016635392, 
+    'before_init_mem_cpu': 5831098368, 
+    'before_init_mem_gpu': 4747206656, 
+    'epoch': 0.04
+}
+```
+
+To compute the total GPU memory allocated for training
+- Total memory delta for train stage = 126487040 bytes ('train_mem_gpu_alloc_delta') + 11016635392 bytes ('train_mem_gpu_peaked_delta')
+- Memory consumption for training = 4747206656 bytes ('before_init_mem_gpu') + 0 bytes ('init_mem_gpu_alloc_delta') + 0 bytes ('init_mem_gpu_peaked_delta') + 126487040 bytes ('train_mem_gpu_alloc_delta') + 11016635392 bytes ('train_mem_gpu_peaked_delta')
+
+#### No Significant Slowdown Using HF Memory Probes 
+| acceleration type         | model_name_or_path                       | num_gpus | batch size | throughput without mem probs (toks/s) | throughput with mem probs (toks/s) | allocated gpu memory (GiB) |
+| ------------------------- | ---------------------------------------- | -------- | ---------- | -------------------------- | ---------------------------------- | -------------------------- |
+| accelerated-peft-bnb      | mistralai/Mistral-7B-v0.1                | 1        | 4          | 3385                       | 3451                               | 15.9                       |
```

Comment:
summarize for the reader, this is too many numbers and the reader will get confused.

Reply:
i suggest to have two concise sections
1. on the slowdown
2. on the differences between nvidia and HF probes.